### DISCUSSION TITLE
implement create new Lakehouse in our workload

### DISCUSSION THREAD

**Discussion by box-turtle** [microsoft/Microsoft-Fabric-workload-development-sample#42] [Help] [answered]
We have a Lakehouse explorer component and want to implement "Add a lakehouse" functionality.

![create-lakehouse](https://github.com/user-attachments/assets/bcf92bfa-6982-43d5-b48e-60d8bc20a18f)

Can we call the same API that is being called above? Will this allow users from other companies to create a lakehouse in their subscription?

```
https://wabi-us-north-central-j-primary-redirect.analysis.windows.net/metadata/workspaces/add277b2-6fd2-4664-b967-7a8968b055a4/artifacts
```

**Comment by ramizsha** [comment, marked as answer]
Hi @box-turtle,
No, you can't use the same API.
Please refer to the public API instead: https://learn.microsoft.com/en-us/rest/api/fabric/lakehouse/items/create-lakehouse?tabs=HTTP

**Reply by tjroberts** [reply]
@ramizsha  We tried using the endpoint in the docs but do not have the right type of permissions. Can you provide an example on how we can generate a token using the app registration? What permissions does it require? 

We tried to use the service principle and the token we have been using previously for other operations we get this error: `PrincipalTypeNotSupported - The operation is not supported for the principal type`

**Reply by ramizsha** [reply]
@tjroberts You can see the required delegated permissions in the doc mentioned above: Lakehouse.ReadWrite.All or Item.ReadWrite.All.
You can add one of those scopes as a dependency on your application in the API permissions tab.
Next, you will need the user to consent to that scope, you can use acquireAccessToken JS API to get a consent window.
To generate a token, you can use OBO flow to exchange the token you get from data plane APIs or control plane APIs.
You can see examples of such flows in the code where we use "GetAccessTokenOnBehalfOf".

**Reply by box-turtle** [reply]
@ramizsha I think we may have some confusion around calling Azure storage apis vs Fabric Lakehouse APIs.

is there a difference in the OBO token creation for a call to https://onelake.dfs.fabric.microsoft.com vs https://api.fabric.microsoft.com

**Reply by box-turtle** [reply]
Our App registration does have Lakehouse.ReadWrite.All and Item.ReadWrite.All added in the API permissions tab as per the documentation.

But the OBO token we are creating that works for our list tables API does not work for the Fabric create table API.

**Reply by ramizsha** [reply]
> @ramizsha I think we may have some confusion around calling Azure storage apis vs Fabric Lakehouse APIs.
> 
> is there a difference in the OBO token creation for a call to https://onelake.dfs.fabric.microsoft.com vs https://api.fabric.microsoft.com

Yes - when calling Onelake (https://onelake.dfs.fabric.microsoft.com), the token generated should be for Azure storage, while when calling Fabric, the token generated should be for Fabric.
See https://github.com/microsoft/Microsoft-Fabric-workload-development-sample/blob/main/Backend/src/Constants/EnvironmentConstants.cs (FabricBackendResourceId & OneLakeResourceId )

**Reply by ramizsha** [reply]
> Our App registration does have Lakehouse.ReadWrite.All and Item.ReadWrite.All added in the API permissions tab as per the documentation.
> 
> But the OBO token we are creating that works for our list tables API does not work for the Fabric create table API.

Can you share what you are using as:
1. Scope
2. ResourceId (audience)
3. API url

**Reply by kedar-lanka** [reply]
@ramizsha can you confirm if the token generated based on below can be used ?

``
curl "https://login.microsoftonline.com/xxx-xxx-xxxxx/oauth2/token" -H "Content-Type: application/x-www-form-urlencoded" --data "grant_type=client_credentials&client_id=xxxxx&client_secret=xxxxx&resource=https%3A%2F%2Fapi.fabric.microsoft.com&scope=https://analysis.windows.net/powerbi/api/.default"
``

the `client_id` that is used in the above url has access to the scopes that you mentioned
<img width="200" alt="Screenshot 2024-07-25 at 5 46 07 PM" src="https://github.com/user-attachments/assets/354b60b3-94e4-41a2-b401-5882f303feb5">

this is one of the scopes for the app
`api://localdevinstance/<tenant_id>/Org.WorkloadSample/lee/Lakehouse.ReadWrite.All`

i have also tried to generate and use token generated using python sdk but that gives me `PrincipalTypeNotSupported` error
```
from azure.identity import ClientSecretCredential
credential = ClientSecretCredential(client_id="xxx", client_secret="xxxx", tenant_id="xxx")
token = credential.get_token("https://analysis.windows.net/powerbi/api/.default").token # PrincipalTypeNotSupported - The operation is not supported for the principal type
token = credential.get_token("https://api.fabric.microsoft.com/.default").token # PrincipalTypeNotSupported - The operation is not supported for the principal type
```

**Reply by ramizsha** [reply]
@kedar-lanka To be able to generate a user token, you'll need to exchange the token you received using OBO flow.
Please refer to this doc and use _ConfidentialClientApplication_ if you're working with python: https://learn.microsoft.com/en-us/python/api/msal/msal.application.confidentialclientapplication?view=msal-py-latest

See there the section about _acquire_token_on_behalf_of_.

**Reply by kedar-lanka** [reply]
@ramizsha i tried this (locally also from backend deployed in kubernetes) but when i use the token that is passed from frontend along for confidential client it gives me invalid subject, audience, issuer

from msal import ConfidentialClientApplication
token = "<generated_by_front_end>"
app = ConfidentialClientApplication(client_id="xxx-xxx-xxxx", client_credential={"client_assertion": token})
downstream_api_access_token = app.acquire_token_on_behalf_of(user_assertion=token, scopes=["FabricLakehouse.Read.All"])
currently both frontend and backend are using the same App so the app has permissions.

frontend is generating token same as the [sample here](https://github.com/microsoft/Microsoft-Fabric-workload-development-sample/blob/main/Frontend/src/index.ui.tsx#L12)

**Reply by ramizsha** [reply]
Please note that scopes should include the audience as well, for example:
_downstream_api_access_token = app.acquire_token_on_behalf_of(user_assertion=token, scopes=["https://analysis.windows.net/powerbi/api/Lakehouse.Read.All"])_
You can see the audience & scope for each permission on your application under API permissions (just click on the permission and a side pane should open).

**Reply by tjroberts** [reply]
I followed Kedar's code above with the additional scope and got this result:
```
>>> token = "<generated_by_front_end>"
>>> app = ConfidentialClientApplication(client_id="<client-id>", client_credential={"client_assertion": token})
>>> downstream_api_access_token = app.acquire_token_on_behalf_of(user_assertion=token, scopes=["FabricLakehouse.Read.All", "https://analysis.windows.net/powerbi/api/Lakehouse.Read.All"])
>>> print(downstream_api_access_token)
```
```
{'error': 'invalid_client', 'error_description': 'AADSTS70025: Client application has no configured federated identity credentials. Trace ID: 7b44057e-dcc3-42bd-8435-fe8b8c959c00 Correlation ID: 024f3923-f3b0-4f45-b941-69d4fa7b954f Timestamp: 2024-07-25 23:18:46Z', 'error_codes': [70025], 'timestamp': '2024-07-25 23:18:46Z', 'trace_id': '7b44057e-dcc3-42bd-8435-fe8b8c959c00', 'correlation_id': '024f3923-f3b0-4f45-b941-69d4fa7b954f'}
```

**Reply by tjroberts** [reply]
@ramizsha Can you please provide a full example of how this token can be generated?

**Reply by ramizsha** [reply]
@tjroberts You did not provide the client secret when initializing the ConfidentialClientApplication.
Also, the first scope you provided "FabricLakehouse.Read.All" is misisng an audience, in that case the audience is considered to be graph.
See example of initializing (in c#) here: https://github.com/microsoft/Microsoft-Fabric-workload-development-sample/blob/main/Backend/src/Program.cs (AddMsalConfidentialClientApplication).
And see example of OBO flow here: https://github.com/microsoft/Microsoft-Fabric-workload-development-sample/blob/main/Backend/src/Controllers/LakeHouseController.cs (_authenticationService.GetAccessTokenOnBehalfOf).
Please note that using a secret is only recommended when developing the workload, In production however, certificates are recommended as they're more secure than client secrets.
